### ISSUE TITLE
Resolve build dependencies separately from others (and take installed packages into account)

### ISSUE THREAD

**Issue by mathstuf** [LLNL/spack#839]
Here is an outline of a test case for some corner cases for using dependency types (#378) to allow build-only dependencies to be resolved separately from other, unrelated, instances of the package in the DAG. The use case is for something like a package `A` which uses `scons` to build which brings in, say `python@2`. Another package, `X` depends on `A` and `B`, but `B` depends on `python@3`. Currently, the DAG will fail to resolve due to the `python` incompatibilities even though they do not strictly conflict.

@tgamblin @dhandeo @eschnett @citibeth 

---

Suppose we are looking to install a package `R`. Here is the dependency graph for `R`:

![spack-deps](https://cloud.githubusercontent.com/assets/97253/14824525/56e0587e-0ba4-11e6-816d-ac4203d7a6aa.png)

First, when traversing the DAG, we get this information:

```
Name    Type    Parent
----    ----    ------
R       b,l,r
D₁      b,l     R
B₁~X    b       D₁
L₁      b,l     D₁
B₁      b       L₁
L₂      b,l     L₁
D₂      b,l     R
B₁+X    b       D₂
```

When traversing, we add specs to the resolution map `M` unless we find specs which match `Type = {b} ∧ Parent ≠ R`. For each such case, we make a new resolution map `Mₖ` and add the non-root traversal over its parent for `{link, run}` dependencies over `D₁`'s `{build, link}` dependencies. This query creates a map for all specs which will need to be loaded during `D₁`'s build. Here, this equates to `{L₁, B₁, B₁~X, L₂}`. There will be another resolution multimap for the `B₁+X` dependency of `D₂` containing `{B₁+X}`.

Here is a snapshot of the dependency resolution maps after the initial traversal:

```
M: {D₁, L₁, B₁, D₂, L₂}
M₁: {L₁, B₁, B₁~X, L₂}
M₂: {B₁+X}
```

Now we must resolve each `Mₖ` against `M`. However, we must choose an order in which to do so. I'm not sure of a fool-proof way to do this, but a sensible one seems to be to sort the maps based on how many entries are contained in the map which are shared with `M` and have multiple specs with that `Mₖ`. Here is the reasoning as to why such a sort is necessary:

Suppose we choose to resolve `M₂` against `M` first. This would force `B₁` from `L₁` to have the `+X` feature. When we then go to resolve `M₁` itself, we will get a conflict between `D₁`'s build dependency of `B₁~X` and the `B₁` required by `L₁` and the resolution fails (which is the case as it is now, but the problem is that there is a solution).

Whatever heuristic we end up using, it has to at least do this (we may also try backtracking to try other sort orders, but this is `O(n!)` in the number of secondary resolution maps created, so avoiding that is best (there are ways that can probably be used to determine that one map can't be resolved after another[1], but that isn't going to help that much).

So in resolving `M₁` against `M`, we determine that `B₁` in `M` must be `~X`. When we resolve `M₂`, there is no internal problems, but it cannot use the same `B₁` as the main build, so it is possible to just build a second `B₁` and use it just for `D₂`'s build.

[1]Namely when the second resolution fails, the second cannot follow the first. If any first resolution fails, there is no solution.

---

As a further enhancement, we can, before merging in any `Mₖ`, resolve `M` against the set of all installed packages. If any dependency is satisfied there, we can ignore any `Mₖ` which is made for that dependency (since we do not need to build it ourselves, none of its build dependencies can "spoil" our main package's dependency resolution). However, we must be able to successfully resolve all such installed dependencies together. I'm not sure what the failure mode of such a resolution should be. Do we kick out certain specs? If so, which ones? Kick out all such dependencies (seems harsh)? Fail with a message that `X` and `Y` conflict, so either specify `^X*<X_hash>`, `^Y*<Y_hash>`, or at least one of `^X*new` and `^Y*new` (the `*` sigil meaning "look for an installed hash" with `new` meaning *don't* look for an installed hash)? It gets messy quickly.

**Comment by mathstuf** [comment]
A clarification for installed package resolution: the first resolution against `M` would be for only those packages which have an associated `Mₖ` (to avoid considering that map at all). A final pass could be done after all possible merging has been done against each remaining `M` to skip building those dependencies.

**Comment by citibeth** [comment]
I would request that any feature resolving against installed packages
(other than spack find and spack location) be optional, and can be turned
off through a configuration in the .spack/ directory.  I am a strong
believer that these things should be accomplishable in other ways (for
example, by completing and improving the functionality of packages.yaml).

I don't like installed package resolution because resolving against
installed packages creates unrepeatable effects.  I want to know that the
series of "spack install" commands I've written up and placed on my website
as installation instructions will work when someone else repeats them on a
clean Spack repo.

Already, we see this issue with "spack location".  It's fine as a one-off
command.  But when people try to base builds around it, a "spack location"
command that worked just fine at first breaks when the set of installed
packages changes (i.e. if you've built with more than one compiler).  This
already causes problems when coming up with a set of "spack location"
commands for one's .bashrc file.

**Comment by mathstuf** [comment]
Installed package resolution against this stuff is a separate can of worms; the first implementation probably won't look at the install tree at all (but since it has been asked for, where it should hook into this stuff has been considered).

**Comment by mathstuf** [comment]
Oh, @adamjstewart was the one interested in the using-installed-packages bit here.

**Comment by tgamblin** [comment]
@citibeth: when installed package resolution is added, there will also be an option on `install` to disable it and to revert to clean-slate concretization.

Also, #360 is getting merged very very very soon, and with that you will be able to refer to a build by a hash.  So you could say:

    `spack load /1234abc`

And you would get a *very* specific build.  If you supply the full hash, there isn't any chance for ambiguity.

**Comment by davydden** [comment]
@mathstuf what is the envisioned resolution for the following case (reworded from [my post in the build dependency PR](https://github.com/LLNL/spack/pull/378#issuecomment-231950398)). Say in addition to the graph above you have: 
```
B_1~X  [link] -> A1~Y
B_1+X  [link] -> A1+Y
B_1    [link] -> A1
R      [link] -> A1+Z
```
What happens to those `A1`'s when we do `spack install R ^A1%xyz+abc`? I would strongly suggest that whatever extra parameters are given in CLI for `A1` (i.e. `^A1%xyz+abc`) shall affect only `R [link] -> A1+Z`.

How to go about those other `A1`'s is another issue. I see two possible solutions.

(1) Let these subtrees (`B_1~X`, `B_1+X`, `B_1`) be built by default without any interplay between them and the main package and couple it with re-usage idea of **build-only** dependencies. 
To clarify, 
```
spack install R ^A1%xyz+abc
```
will produce results indistinguishable from 
```
spack install B_1+X [whatever extra parameters]
spack install B_1~X [whatever other extra parameters]
spack install R ^A1%xyz+abc  # reuse B_1+X and B_1~X  from already installed packages.
```

(2) Develop a complicated CLI to specify how to build `A1` dependencies of `B_1~X`, `B_1+X`, `B_1` and `R` with different compiler options or extra variants: 
```
spack install R ^A1%xyz (^B_1~X^A1%abc) (^B_1+X^A1%cde) (^B_1^A1%fgh)
```
If no extra options are given, build defaults.

**Comment by citibeth** [comment]
On Apr 28, 2016 2:01 PM, "Todd Gamblin" <notifications@github.com> wrote:
>
> @citibeth: when installed package resolution is added, there will also be
an option on install to disable it and to revert to clean-slate
concretization.

There needs to be a config option so I don't always have to say it on the
cmd line.  We might find that adding variant defaults to packages. Py
obviates the desire for this pr.

**Comment by mathstuf** [comment]
> an option on install to disable it and to revert to clean-slate concretization

Yeah, that's policy stuff (for discussion when there's actual code for an option to even affect). This issue is about how to even do it in the first place.

> We might find that adding variant defaults to packages.py obviates the desire for this pr.

Default (or even propagated) variants won't solve this issue at all.

**Comment by mathstuf** [comment]
> What happens to those A1's when we do spack install R ^A1%xyz+abc? I would strongly suggest that whatever extra parameters are given in CLI for A1 (i.e. ^A1%xyz+abc) shall affect only R [link] -> A1+Z.

This would be a possible solution, but if you are starting from scratch, not have it affect the `A₁` builds you'd need anyways? Once code exists, this should be included in the test cases.

How to go about those other A1's is another issue. I see two possible solutions.

> (1) Let these subtrees (B_1~X, B_1+X, B_1) be built by default without any interplay between them and the main package and couple it with re-usage idea of build-only dependencies.

Such a solution is probably a good first-pass at it. The hard part is finding specs you can combine together to minimize build time.

> spack install R ^A1%xyz+abc
> 
> will produce results indistinguishable from
> 
> spack install B_1+X [whatever extra parameters]
> spack install B_1~X [whatever other extra parameters]
> spack install R ^A1%xyz+abc  # reuse B_1+X and B_1~X  from already installed packages.

The "use already built packages where possible" is the second phase of the issue. Currently we take the whole DAG into account when doing concretization and only once that is done care about what is installed.

> (2) Develop a complicated CLI to specify how to build A1 dependencies of B_1~X, B_1+X, B_1 and R with different compiler options or extra variants:
> 
> spack install R ^A1%xyz (^B_1~X^A1%abc) (^B_1+X^A1%cde) (^B_1^A1%fgh)

Is there really a use case where one cares about what toolset a build dependency is compiled with to such an extent? If you really want to use the installed ones, do `B₁/hashofinstall`.

**Comment by davydden** [comment]
> This would be a possible solution, but if you are starting from scratch, why not have it affect the A₁ builds you'd need anyways?

My original issue was with cases like
```
dealii [link] -> petsc [build] -> python [link]-> zlib
   \
    \ [link] -> zlib
```
where, say, making `spack install dealii%gcc ^petsc ^python%clang` would not allow me to have the whole python thing be built with `clang` whereas having `zlib` in deal.II to be built with `gcc`. If we agree that we don't care about how build-only dependencies are built, then i would say it makes sense that CLI input for concretization (be it compiler of root package `dealii%gcc` or specifics of dependencies `dealii ^zlib%gcc`) does not propagate through `build` nodes.

> Is there really a use case where one cares about what toolset a build dependency is compiled with to such an extent?

i would say -- no. So i would be perfectly happy with (i) solution.

**Comment by davydden** [comment]
> > We might find that adding variant defaults to packages.py obviates the desire for this pr.

> Default (or even propagated) variants won't solve this issue at all.

:+1: This PR is really a mast have, sooner or later. Dependency types is just a first big step into this direction.

**Comment by tgamblin** [comment]
@mathstuf @scheibelp since you guys are on the same page per #2537, do you want to take a stab at this together?

**Comment by citibeth** [comment]
This issue actually advocates for two things in the title; which one are @mathstuf and @scheibelp looking into at this point?  I believe that resolving dependencies based on already-installed stuff (or other such lists of specs) needs the upcoming new concretizer?

**Comment by tgamblin** [comment]
Specifically fixing things for the build dependencies.  I think this should probably be two issues -- considering what is installed is harder and would need more of a rework.  @mathstuf is actually already working on that, but it's prototype at this point.

**Comment by citibeth** [comment]
> @mathstuf <https://github.com/mathstuf> is actually already working on
that, but it's prototype at this point.

OK... when it's ready, I will definitely want to see the following features:
 1. List already-built stuff into a user-editable file
 2. Concretize based on a file listing already-built stuff
 3. For convenience, combine (1) and (2) together.

**Comment by tgamblin** [comment]
Why is that not encompassed by `packages.yaml` + the install database?  Are you saying you want a per-project context of already built stuff, which is why you need the file?  What is the use case for this?

**Comment by scheibelp** [comment]
> @mathstuf @scheibelp since you guys are on the same page per #2537, do you want to take a stab at this together?

I am looking into this at the moment. If @mathstuf you have some free time I'd appreciate doing something more collaborative. IIRC you mentioned somewhere you do not though, in which case I'll work on a solution and then if possible your input would be quite valuable.

**Comment by citibeth** [comment]
>
>
> Why is that not encompassed by packages.yaml + the install database? Are
> you saying you want a per-project context of already built stuff, which is
> why you need the file? What is the use case for this?
>

Installing with respect on "whatever happens to be in my Spack" is not
repeatable.  Installing with respect to the contents of a file IS
repeatable.  There are many uses cases here:

1. I'm having weird problems and I ask you for help.  You run the same
installation and have no problems at all.  After a lot of back-and-forth
and heartache, we figure out my problems are because Spack is picking up
some package I have installed that you don't.  This would have been SO much
easier if I'd sent you the file listing my Spack installations to begin
with.  Especially if the problems are apparent in `spack spec` and don't
require the packages to ACTUALLY be installed.

2. I fiddle around and finally get an installation that works.  Then I
clear out my Spack, re-install, and discover that I have different versions
of stuff.  Because I installed them in different orders, etc.  Having a
single file of "this is what I had installed before, and I want you to
respect it" will help me in building stuff in my new repo.

3. I like what I've installed, except for some parts of it.  So I list out
the file, delete the parts I don't like, and then do installs with respect
to that file on every node of our cluster.

Is this like a per-project `packages.yaml`?  Sort of.  Except this is a
list that is equivalent to the full specs of a bunch of installed things.
My guess is that the constraints added by `packages.yaml` are a little
different.

**Comment by tgamblin** [comment]
Ok right.  We've always said we'll keep a deterministic/stateless/repeatable option.  Just making sure this is not something new.

**Comment by hartzell** [comment]
+1 for something like what @citibeth is asking for, as it sounds like it would be useful for my "set up our standard set of apps on a new system" use case.

Right now I have a script full of `spack install this` statements and I awakened the discussion about meta-packages in the past to scratch the same itch.

**Comment by tgamblin** [comment]
@hartzell: can you attach your script?

**Comment by citibeth** [comment]
I used to use scripts like @hartzell too.  I now use "dummy packages" (meta-packages), which have their plusses and minuses compared to the old scripts.

**Comment by hartzell** [comment]
@tgamblin -- Here's the most recent incantation.  It's cribbed from something @eschnett [published to the google group](https://groups.google.com/forum/#!topic/spack/h_UtM17jV7Y). Calling it a "script" is a bit presumptuous....

The compiler bit is particularly sub-par.  The first time it runs it uses the system compiler to build me a gcc@5.4.0. If you run it again, it uses gcc@5.4.0 to build a gcc@5.4.0.  I don't care, it's easy enough to comment out, but the next time I find a spare TUIT lying around I'll check whether spack has a gcc@5.4.0 installed before blindly building another one.  I almost always run it on a clean instance so it Works For Me.

```sh
# Compiler
spack install gcc@5.4.0

# This might not do what you want if there's already
# a 5.4.0 on the system....
# Now all of the things will be built with %gcc@5.4.0
spack compiler add --scope site $(spack location -i gcc@5.4.0)

# R things
spack install R@3.3.1+X ^cairo+X^pango+X
spack install r-devtools@1.11.1 ^cairo+X^pango+X
spack install r-shiny@0.13.2 ^cairo+X^pango+X
spack install r-xml@3.98-1.5 ^cairo+X^pango+X

# python things
spack install python@2.7.12
spack install py-flake8@3.0.4
# what other packages to we want to install?

spack install ack@2.14
spack install ant@1.9.7
spack install bcftools@1.3.1
spack install boost@1.62.0
spack install bwa@0.7.15
spack install cask@0.8.1
spack install cmake@3.6.1
spack install emacs@24.5
spack install emacs@25.1+X
spack install fastx_toolkit@0.0.14
spack install git@2.9.3
spack install go@1.7.4
spack install hdf5@1.8.15
spack install htop@2.0.2
spack install htslib@1.3.1
spack install jdk@8u92-linux-x64
spack install lapack@3.6.1
spack install lmod@6.4.5
spack install maven@3.3.9
spack install nextflow@0.20.1
spack install openblas@0.2.19
spack install parallel@20160422
spack install perl@5.22.2
spack install perl@5.24.0
spack install prank@150803
spack install samtools@1.3.1
spack install seqtk@1.2
spack install swig@3.0.10
spack install texlive@live
spack install tmux@2.3
spack install tree@1.7.0
spack install unison@2.48.4
spack install valgrind@3.11.0
spack install xauth@1.0.9
```

**Comment by hartzell** [comment]
> I used to use scripts like @hartzell too

I may someday be able to say the same thing.

After @citibeth's explanation of her meta-packages in #2049 and link to [env-modele](https://github.com/citibeth/spack/blob/efischer/develop/var/spack/repos/builtin/packages/env-modele/package.py) I goofed around with them a bit.    Just Need More TUITS....  (that should be a bumper sticker or a t-shirt or ...).

**Comment by adamjstewart** [comment]
@tgamblin The script I use to install one of my common stacks looks like:
```
compilers=(
    %gcc
    %intel
    %pgi
    %nag
)

mpis=(
    openmpi+psm~verbs
    openmpi~psm+verbs
    mvapich2+psm~mrail
    mvapich2~psm+mrail
    mpich+verbs
)

for compiler in "${compilers[@]}"
do
    # Serial installs
    spack install szip           $compiler
    spack install hdf            $compiler
    spack install hdf5           $compiler
    spack install netcdf         $compiler
    spack install netcdf-fortran $compiler
    spack install ncview         $compiler
    spack install nco            $compiler

    # Parallel installs
    for mpi in "${mpis[@]}"
    do
        spack install $mpi            $compiler
        spack install hdf5~cxx+mpi    $compiler ^$mpi
        spack install h5hut+mpi       $compiler ^$mpi
        spack install parallel-netcdf $compiler ^$mpi
    done
done
```
I think it would be good to store a list of what is installed for the sake of being able to reinstall the same stack on a different Spack installation every time the hash changes.

**Comment by alalazo** [comment]
Closing as outdated. This issue has been linked into #38225
